test/UpgradeRegressionManager.invariant.t.sol

```diff
@@ -0,0 +1,122 @@
+import {UpgradeRegressionManager} from "src/contracts/UpgradeRegressionManager.sol";
+import {MockTimelockTarget} from "test/mocks/MockTimelockTarget.sol";
+import {SimpleContract} from "test/helpers/SimpleContract.sol";
+import {UpgradeRegressionManagerHandler} from "test/handlers/UpgradeRegressionManager.handler.sol";
+import {ITimelockTarget} from "src/interfaces/ITimelockTarget.sol";
+
+contract UpgradeRegressionManagerInvariantTest is Test {
+  UpgradeRegressionManager public urm;
+  MockTimelockTarget public timelockTarget;
+  SimpleContract public target;
+  UpgradeRegressionManagerHandler public handler;
+
+  address public admin = makeAddr("admin");
+  address public guardian = makeAddr("guardian");
+
+  function setUp() public {
+    timelockTarget = new MockTimelockTarget();
+    timelockTarget.setDelay(5 seconds);
+
+    target = new SimpleContract();
+
+    urm = new UpgradeRegressionManager(ITimelockTarget(timelockTarget), admin, guardian, 5 seconds);
+
+    handler = new UpgradeRegressionManagerHandler(urm, admin, guardian, target);
+
+    // target the handler for invariant testing
+    targetContract(address(handler));
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Rollback State Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice A rollback cannot exist in both `rollbackQueueExpiresAt` and `rollbackExecutableAt` simultaneously
+  /// @dev Rationale: Ensures rollback is in exactly one state at a time
+  function invariant_rollbackStateMutualExclusion() public view {
+    uint256 rollbackId = handler.getRollbackId();
+
+    bool inQueueingState = urm.rollbackQueueExpiresAt(rollbackId) != 0;
+    bool inExecutionState = urm.rollbackExecutableAt(rollbackId) != 0;
+
+    assertFalse(inQueueingState && inExecutionState, "Rollback cannot exist in both queueing and execution state");
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Timing Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice `rollbackQueueWindow` must always be greater than 0
+  /// @dev Ensures valid queue window for rollbacks
+  function invariant_rollbackQueueWindowPositive() public view {
+    uint256 _rollbackQueueWindow = urm.rollbackQueueWindow();
+    assertGt(_rollbackQueueWindow, 0, "Rollback queue window must be greater than 0");
+  }
+
+  /// @notice When a rollback is queued, its ETA must be in the future
+  /// @dev Ensures proper timing for execution
+  function invariant_etaInFutureAfterQueue() public view {
+    uint256 _rollbackId = handler.getRollbackId();
+    uint256 _eta = urm.rollbackExecutableAt(_rollbackId);
+    uint256 _currentTime = block.timestamp;
+
+    // nothing to test — either not proposed or still within queue window
+    if (_eta == 0) {
+      return;
+    }
+
+    assertGt(_eta, _currentTime, "Rollback ETA must be in the future");
```

Comment:
Two thoughts:

1. Without a handler to warp into the future, I'm not sure we're ever testing anything.
2. Is this true? A proposal can be past it's ETA but has not been executed yet, right?